### PR TITLE
WIP: first steps towards multiple lanes and lane changes

### PR DESCRIPTION
**Pull request by HenKlei**
We should start a discussion on what to change how by using the current version of the new lane_change branch as an example (not as the final way to go, that's for sure). I am not quite sure if this really goes into the right direction so we should have a deeper look on how to manage the different vehicles and their relationships. Please let me know if you have any ideas on how to simplify the approach, at the moment that all seems to be a bit inflated compared to what we actually want to do.

Suggestions on how to treat the circular shape of our road (as for a vehicle the predecessor on the right lane could be the same as the successor on the right lane) are greatly appreciated. From my point of view using a circle instead of a straight road could be one of the main reason for the problems, difficulties and exceptional cases we face here.

### CODE REVIEW

**Comment by HenKlei** [comment]
Unfortunately I will not be able to take part in the meeting on Wednesday. Maybe, if you are talking about this topic there, you can sketch your ideas in a comment here so that I can stay up to date. That would be nice, thank you in advance.

File: trafficFlow/carFollowingModel/carFollowingModel.py [hunk 1]
```diff
@@ -39,4 +45,8 @@ def simulate_one_step(self, time_discretization_scheme, t, dt):
         y = time_discretization_scheme.apply(t, dt, y)
         for i, vehicle in enumerate(self.road.get_vehicles()):
             vehicle.position = self.road.get_position(y[i], vehicle.lane)
-            vehicle.velocity = y[self.road.get_number_of_vehicles() + i]
+            if y[self.road.get_number_of_vehicles() + i] < 0.:
+                vehicle.velocity = 0.
+            else:
+                vehicle.velocity = y[self.road.get_number_of_vehicles() + i]
```
Comment (tobiasleibner) [inline trafficFlow/carFollowingModel/carFollowingModel.py:51]: ``y`` should be a ``numpy`` array, then you can replace this by something like
```python3
num_vehicles = self.road.get_number_of_vehicles()
positions = y[:num_vehicles]
velocities = y[num_vehicles:]
# set negative velocites to zero
velocities[velocities < 0.] = 0.
# update vehicles
for i, vehicle in enumerate(self.road.get_vehicles()):
  vehicle.position, vehicle.velocity = positions[i], velocities[i]
```

File: trafficFlow/carFollowingModel/drivers/baseDriver.py [hunk 2]
```diff
@@ -34,22 +37,243 @@ class BaseDriver:
...
+        if self.predecessor_left:
+            return self.velocity - self.predecessor_left.velocity
+        else:
+            return -math.inf
+
+    def get_speed_difference_to_successor_left(self):
+        if self.successor_left:
+            return self.successor_left.velocity - self.velocity
+        else:
+            return math.inf
```
Comment (tobiasleibner) [inline trafficFlow/carFollowingModel/drivers/baseDriver.py:133]: This is a lot of code duplication, you should write a single ``get_distance`` and ``get_speed_difference`` function that takes a vehicle and pass ``self.predecessor_left``, ``self.successor_right`` etc. to these functions.

File: trafficFlow/carFollowingModel/lanes/baseLane.py [hunk 3]
```diff
@@ -21,16 +21,23 @@ class BaseLane:
         place the vehicles in a default manner on the lane
     """
 
-    def __init__(self):
+    def __init__(self, full_length=1000.):
+        self.full_length = full_length
         self.number_of_vehicles = 0
```
Comment (tobiasleibner) [inline trafficFlow/carFollowingModel/lanes/baseLane.py:26]: No need to explicitly store ``number_of_vehicles``, you can just use ``len(vehicles)``.

**Review by tobiasleibner — COMMENTED** [review]
Yep, this really gets too complex, there should be a way to simplify this. To start with, I would keep the ``vehicles`` list in each lane ordered, so instead of appending new vehicles at the end, I would insert each vehicle according to its ``position``. That way, you don't have to keep track of predecessors and successors (in the same lane) and changing a lane is just removing the vehicle from the list in the old lane and inserting it at the correct position in the new lane.
Keeping track of predecessor and successor in the neighbor lanes probably improves performance, but at the cost of adding a lot of complexity. Maybe you should start with computing these on the fly, using a method like
```python3
successor, predecessor = lane.nearby_vehicles(vehicle.position)
```
As long as there are only a few vehicles, this should still be very fast. If we want to simulate hundreds of vehicles in the future, this might be too slow, but we can still add a more complex implementation then if really needed.

File: trafficFlow/carFollowingModel/drivers/baseDriver.py (hunk 2 above)
Comment (tobiasleibner) [reply trafficFlow/carFollowingModel/drivers/baseDriver.py:133]: If the implementation is changed such that  the ``vehicles`` list is ordered, the driver doesn't have to keep track of any neighbors. Instead, all required neighbors should be provided to the ``vehicle`` by the ``lane``.

**Comment by HenKlei** [comment]
This seems to be a very good idea. When the lists are sorted it is also much easier to remove cars, for example if they leave the road or so. When doing it right it should also be possible to introduce crossings or ramps using sorted lists and without keeping track of predecessors and so on. I think many of the methods introduced so far become obsolete when changing to sorted lists so that we get rid of much of the redundant and inefficient code.

**Comment by HenKlei** [comment]
For these changes I will start from the current stable state on master and open a new branch. I will try to commit and push more frequently so that we can check together whether this is going in the right direction.

**Comment by HenKlei** [comment]
I just added a new branch where some of the changes discussed here are introduced. In particular, I tried to make the units consistent and now the vehicles seem to behave in a physically meaningful way. It is also possible to add multiple lanes, but the vehicles can not change the lanes so far. I'm going to open a pull request so that we can discuss the new version.

**Pull request closed without merging**